---
title: "Comparison of characteristic features across count data sets"
author: ""
date: ""
output: 
  html_document:
    toc: true
  pdf_document:
    toc: true
always_allow_html: yes
references:
- id: Love2014DESeq2
  title: Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2
  author:
  - family: Love
    given: Michael I
  - family: Huber
    given: Wolfgang
  - family: Anders
    given: Simon
  container-title: Genome Biology
  volume: 15
  page: 550
  type: article-journal
  URL: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8
  issued:
    year: 2014
- id: Robinson2010edgeR
  title: edgeR-a Bioconductor package for differential expression analysis of digital gene expression data
  author:
  - family: Robinson
    given: Mark D
  - family: McCarthy
    given: Davis J
  - family: Smyth
    given: Gordon K
  container-title: Bioinformatics
  volume: 26
  page: 139-140
  type: article-journal
  URL: https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btp616
  issued:
    year: 2010
- id: Robinson2010TMM
  title: A scaling normalization method for differential expression analysis of RNA-seq data
  author:
  - family: Robinson
    given: Mark D
  - family: Oshlack
    given: Alicia
  container-title: Genome Biology
  volume: 11
  page: R25
  type: article-journal
  URL: https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25
  issued:
    year: 2010
---

## Description

`r I(description)`

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = showCode, warning = FALSE, message = FALSE)

# suppressPackageStartupMessages(library(DESeq2))
# suppressPackageStartupMessages(library(edgeR))
# suppressPackageStartupMessages(library(dplyr))
# suppressPackageStartupMessages(library(tidyr))
# suppressPackageStartupMessages(library(ggplot2))

## Determine number of rows and columns in facetted plots 
## (first element = columns, second element = rows)
colRow <- grDevices::n2mfrow(length(ddsList))

## Set height/width of each panel in multi-panel plots
panelSize <- 4

plots <- list()

thm <- theme_bw() + 
  theme(axis.text = element_text(size = 15),
        axis.title = element_text(size = 14),
        strip.text = element_text(size = 15))

## Check whether the emo package is available (not currently used)
has_emo <- suppressPackageStartupMessages(suppressWarnings(require("emo") == TRUE))
```

```{r}
## Help function to make data table with K-S statistics and p-values
makeDF <- function(df, column) {
  if (length(unique(df$dataset)) > 1) {
    ## Generate data frame with all pairs of data sets
    ks_df <- as.data.frame(t(combn(unique(df$dataset), 2))) %>%
      dplyr::rename(dataset1 = V1, dataset2 = V2)
    ## Calculate K-S statistic and p-value for each pair of data sets
    ks_res <- as.data.frame(t(sapply(1:nrow(ks_df), function(i) {
      kst <- ks.test(df[, column][df$dataset == ks_df$dataset1[i]], 
                     df[, column][df$dataset == ks_df$dataset2[i]])
      c(statistic = round(kst$statistic, 3), pvalue = round(kst$p.value, 3))
    })))
    ks_df$`K-S statistic` <- ks_res$statistic
    ks_df$`K-S p-value` <- ks_res$pvalue
    DT::datatable(ks_df, options = list(
      columnDefs = list(list(className = "dt-center", targets = 3:ncol(ks_df)))
    ))
  } 
}

## Help function to determine the number of objects to calculate the pairwise
## correlations for
getNForCorr <- function(x, Nmax) {
  (min(x, Nmax) * (min(x, Nmax) - 1)) / 2
}
```


```{r, message = FALSE}
message("Calculating dispersions. This can take some time...")
obj <- lapply(ddsList, function(ds) {
  ## --- edgeR
  dge <- edgeR::DGEList(counts = counts(ds))
  des <- model.matrix(design(ds), data = SummarizedExperiment::colData(ds))
  dge <- edgeR::calcNormFactors(dge)
  dge <- edgeR::estimateDisp(dge, design = des)
  ## --- DESeq2
  dds <- DESeq2::estimateSizeFactors(ds, type = "poscounts")
  dds <- DESeq2::estimateDispersions(dds, quiet = TRUE)
  list(dge = dge, dds = dds)
})
```

```{r, message = FALSE}
## Calculate between-sample correlations
message("Calculating sample correlations.")
sampleCorrDF <- do.call(rbind, lapply(obj, function(x) {
  cpms <- edgeR::cpm(x$dge, prior.count = 2, log = TRUE)
  if (ncol(cpms) > subsampleSize) {
    set.seed(123)
    cpms <- cpms[, sample(seq_len(ncol(cpms)), subsampleSize, replace = FALSE)]
  }
  corrs <- cor(cpms, use = "pairwise.complete.obs", method = "spearman")
  data.frame(
    Correlation = corrs[upper.tri(corrs)]
  )
})) %>% dplyr::mutate(dataset = rep(names(obj), sapply(obj, function(x) getNForCorr(ncol(x$dge), subsampleSize))))
```

```{r, message = FALSE}
## Calculate between-variable correlations
message("Calculating variable correlations.")
featureCorrDF <- do.call(rbind, lapply(obj, function(x) {
  cpms <- edgeR::cpm(x$dge, prior.count = 2, log = TRUE)
  cpms <- cpms[genefilter::rowVars(cpms) > 0, ]
  if (nrow(cpms) > subsampleSize) {
    set.seed(123)
    cpms <- cpms[sample(seq_len(nrow(cpms)), subsampleSize, replace = FALSE), ]
  }
  corrs <- cor(t(cpms), use = "pairwise.complete.obs", method = "spearman")
  data.frame(
    Correlation = corrs[upper.tri(corrs)]
  )
})) %>% dplyr::mutate(dataset = rep(names(obj), sapply(obj, function(x) getNForCorr(nrow(x$dge), subsampleSize))))
```

```{r}
## Summarize sample characteristics
sampleDF <- do.call(rbind, lapply(obj, function(x) {
  data.frame(
    Libsize = colSums(x$dge$counts),
    Fraczero = colMeans(x$dge$counts == 0),
    TMM = x$dge$samples$norm.factors
  )
})) %>% dplyr::mutate(dataset = rep(names(obj), sapply(obj, function(x) ncol(x$dge))))
```

```{r}
## Summarize variable characteristics
featureDF <- do.call(rbind, lapply(obj, function(x) {
  data.frame(
    Tagwise = sqrt(x$dge$tagwise.dispersion),
    Common = sqrt(x$dge$common.dispersion),
    Trend = sqrt(x$dge$trended.dispersion),
    AveLogCPM = x$dge$AveLogCPM,
    average_log2_cpm = apply(edgeR::cpm(x$dge, prior.count = 2, log = TRUE), 1, mean), 
    variance_log2_cpm = apply(edgeR::cpm(x$dge, prior.count = 2, log = TRUE), 1, var),
    Fraczero = rowMeans(x$dge$counts == 0),
    dispGeneEst = rowData(x$dds)$dispGeneEst,
    dispFit = rowData(x$dds)$dispFit,
    dispFinal = rowData(x$dds)$dispersion,
    baseMean = rowData(x$dds)$baseMean
  )
})) %>% dplyr::mutate(dataset = rep(names(obj), sapply(obj, function(x) nrow(x$dge))))
```

```{r}
## Summarize data set characteristics
datasetDF <- do.call(rbind, lapply(obj, function(x) {
  data.frame(
    prior_df = paste0("prior.df = ", round(x$dge$prior.df, 2)),
    nVars = nrow(x$dge$counts),
    nSamples = ncol(x$dge$counts)
  )
})) %>% dplyr::mutate(AveLogCPM = 0.8 * max(featureDF$AveLogCPM)) %>%
  dplyr::mutate(Tagwise = 0.9 * max(featureDF$Tagwise)) %>%
  dplyr::mutate(dataset = names(obj))
```

## Data set dimensions {.tabset .tabset-pills}

The number of samples (columns) and variables (rows) in each data set. 

### Number of samples (columns)

```{r nSamples, fig.width = 10, fig.height = 7}
plots[["nSamples"]] <- 
  ggplot(datasetDF, aes(x = dataset, y = nSamples, fill = dataset)) + 
  geom_bar(stat = "identity", alpha = 0.5) + 
  xlab("") + ylab("Number of samples (columns)") + 
  thm + theme(axis.text.x = element_text(size = 15, angle = 90, hjust = 1, vjust = 0.5))
plots[["nSamples"]]
```

### Number of variables (rows)

```{r nVariables, fig.width = 10, fig.height = 7}
plots[["nVariables"]] <- 
  ggplot(datasetDF, aes(x = dataset, y = nVars, fill = dataset)) + 
  geom_bar(stat = "identity", alpha = 0.5) + 
  xlab("") + ylab("Number of variables (rows)") + 
thm + theme(axis.text.x = element_text(size = 15, angle = 90, hjust = 1, vjust = 0.5))
plots[["nVariables"]]
```

## Dispersion/BCV plots {.tabset .tabset-pills}

The association between the average abundance and the dispersion or "biological 
coefficient of variation" (sqrt(dispersion)), as calculated by
[edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html)
[@Robinson2010edgeR] and
[DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq2.html) 
[@Love2014DESeq2]. In the edgeR plot, the estimate of the prior degrees of
freedom is indicated.

### edgeR

The black dots represent the tagwise dispersion estimates, the red line the 
common dispersion and the blue curve represents the trended dispersion
estimates.

```{r BCVedgeR, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["BCVedgeR"]] <- 
  ggplot(featureDF %>% dplyr::arrange(AveLogCPM), 
         aes(x = AveLogCPM, y = Tagwise)) + geom_point(size = 0.25) + 
  facet_wrap(~dataset, nrow = colRow[2]) + 
  geom_line(aes(y = Trend), color = "blue", size = 1.5) + 
  geom_line(aes(y = Common), color = "red", size = 1.5) +
  geom_text(data = datasetDF, aes(label = prior_df)) + 
  xlab("Average log CPM") + ylab("Biological coefficient of variation") + 
  thm
plots[["BCVedgeR"]]
```

### DESeq2

The black dots are the gene-wise dispersion estimates, the red curve the fitted
mean-dispersion relationship and the blue circles represent the final dispersion
estimates.

```{r dispersionDESeq2, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["dispersionDESeq2"]] <- 
  ggplot(featureDF %>% dplyr::arrange(baseMean), 
         aes(x = baseMean, y = dispGeneEst)) + geom_point(size = 0.25) + 
  facet_wrap(~dataset, nrow = colRow[2]) + scale_x_log10() + scale_y_log10() +  
  geom_point(aes(y = dispFinal), color = "lightblue", shape = 21) + 
  geom_line(aes(y = dispFit), color = "red", size = 1.5) + 
  xlab("Base mean") + ylab("Dispersion") + 
  thm
plots[["dispersionDESeq2"]]
```

## Mean-variance plots {.tabset .tabset-pills}

The difference between these mean-variance plots and the mean-dispersion plots 
above is that the plots in this section do not take the information about the 
experimental design and sample grouping into account, but display the mean and 
variance of log2(CPM) estimates across all samples, calculated using the `cpm`
function from
[edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) 
[@Robinson2010edgeR], with a prior count of 2.

### Separate scatter plots

```{r meanVarSepScatter, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["meanVarSepScatter"]] <- 
  ggplot(featureDF, aes(x = average_log2_cpm, y = variance_log2_cpm)) + 
  geom_point(size = 0.75) + facet_wrap(~dataset, nrow = colRow[2]) + 
  xlab("Mean of log2(CPM)") + ylab("Variance of log2(CPM)") + 
  thm
plots[["meanVarSepScatter"]]
```

### Overlaid scatter plots

```{r meanVarOverlaidScatter}
plots[["meanVarOverlaidScatter"]] <- 
  ggplot(featureDF, aes(x = average_log2_cpm, y = variance_log2_cpm, color = dataset)) + 
  geom_point(size = 0.75, alpha = 0.5) +
  xlab("Mean of log2(CPM)") + ylab("Variance of log2(CPM)") + 
  thm
plots[["meanVarOverlaidScatter"]]
```

## Library sizes {.tabset .tabset-pills}

The distribution of the total count per sample, i.e., the column sums of the
respective data matrices.

### Separate histograms

```{r libsizeSepHist, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["libsizeSepHist"]] <- 
  ggplot(sampleDF, aes(x = Libsize)) + geom_histogram(bins = 30) + 
  facet_wrap(~dataset, nrow = colRow[2]) +
  xlab("Library size") + thm
plots[["libsizeSepHist"]]
```

### Overlaid histograms

```{r libsizeOverlaidHist}
plots[["libsizeOverlaidHist"]] <- 
  ggplot(sampleDF, aes(x = Libsize, group = dataset, fill = dataset)) + 
  geom_histogram(bins = 30, alpha = 0.5, position = "identity") +
  xlab("Library size") + thm
plots[["libsizeOverlaidHist"]]
```

### Density plots

```{r libsizeDensity}
plots[["libsizeDensity"]] <- 
  ggplot(sampleDF, aes(x = Libsize, group = dataset, color = dataset)) + 
  geom_line(alpha = 0.5, size = 1.5, stat = "density") +
  xlab("Library size") + thm
plots[["libsizeDensity"]]
```

### Density plots (filled)

```{r libsizeDensityFilled}
plots[["libsizeDensityFilled"]] <- 
  ggplot(sampleDF, aes(x = Libsize, group = dataset, fill = dataset)) + 
  geom_density(alpha = 0.5) +
  xlab("Library size") + thm
plots[["libsizeDensityFilled"]]
```

### Box plots

```{r libsizeBox}
plots[["libsizeBox"]] <- 
  ggplot(sampleDF, aes(x = dataset, y = Libsize, color = dataset)) + 
  geom_boxplot(outlier.size = -1, alpha = 0.5) + 
  geom_point(position = position_jitter(width = 0.2), size = 2, alpha = 0.5) + 
  ylab("Library size") + xlab("") + thm
plots[["libsizeBox"]]
```

### Violin plots

```{r libsizeViolin}
plots[["libsizeViolin"]] <- 
  ggplot(sampleDF, aes(x = dataset, y = Libsize, fill = dataset)) + 
  geom_violin(alpha = 0.5) + 
  ylab("Library size") + xlab("") + thm
plots[["libsizeViolin"]]
```

### Empirical cumulative distribution function 

```{r libsizeEcdf}
plots[["libsizeEcdf"]] <- 
  ggplot(sampleDF, aes(x = Libsize, group = dataset, color = dataset)) + 
  stat_ecdf(size = 1.5, alpha = 0.5) + ylab("Cumulative proportion") + 
  xlab("Library size") + thm
plots[["libsizeEcdf"]]

makeDF(sampleDF, "Libsize")
```

## TMM normalization factors {.tabset .tabset-pills}

The TMM normalization factors [@Robinson2010TMM], intended to adjust for
differences in RNA composition, as calculated by
[edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html)
[@Robinson2010edgeR].

### Separate histograms

```{r tmmSepHist, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["tmmSepHist"]] <- 
  ggplot(sampleDF, aes(x = TMM)) + geom_histogram(bins = 30) + 
  facet_wrap(~dataset, nrow = colRow[2]) +
  xlab("TMM normalization factor") + thm
plots[["tmmSepHist"]]
```

### Overlaid histograms

```{r tmmOverlaidHist}
plots[["tmmOverlaidHist"]] <- 
  ggplot(sampleDF, aes(x = TMM, group = dataset, fill = dataset)) + 
  geom_histogram(bins = 30, alpha = 0.5, position = "identity") +
  xlab("TMM normalization factor") + thm
plots[["tmmOverlaidHist"]]
```

### Density plots

```{r tmmDensity}
plots[["tmmDensity"]] <- 
  ggplot(sampleDF, aes(x = TMM, group = dataset, color = dataset)) + 
  geom_line(alpha = 0.5, size = 1.5, stat = "density") +
  xlab("TMM normalization factor") + thm
plots[["tmmDensity"]]
```

### Density plots (filled) 

```{r tmmDensityFilled}
plots[["tmmDensityFilled"]] <- 
  ggplot(sampleDF, aes(x = TMM, group = dataset, fill = dataset)) + 
  geom_density(alpha = 0.5) +
  xlab("TMM normalization factor") + thm
plots[["tmmDensityFilled"]]
```

### Box plots

```{r tmmBox}
plots[["tmmBox"]] <- 
  ggplot(sampleDF, aes(x = dataset, y = TMM, color = dataset)) + 
  geom_boxplot(outlier.size = -1, alpha = 0.5) + 
  geom_point(position = position_jitter(width = 0.2), size = 2, alpha = 0.5) + 
  ylab("TMM normalization factor") + xlab("") + thm
plots[["tmmBox"]]
```

### Violin plots

```{r tmmViolin}
plots[["tmmViolin"]] <- 
  ggplot(sampleDF, aes(x = dataset, y = TMM, fill = dataset)) + 
  geom_violin(alpha = 0.5) + 
  ylab("TMM normalization factor") + xlab("") + thm
plots[["tmmViolin"]]
```

### Empirical cumulative distribution function 

```{r tmmEcdf}
plots[["tmmEcdf"]] <- 
  ggplot(sampleDF, aes(x = TMM, group = dataset, color = dataset)) + 
  stat_ecdf(size = 1.5, alpha = 0.5) + ylab("Cumulative proportion") + 
  xlab("TMM normalization factor") + thm
plots[["tmmEcdf"]]

makeDF(sampleDF, "TMM")
```

## Expression distributions (average log CPM) {.tabset .tabset-pills}

The distribution of average abundance values for the variables. 

### Separate histograms

```{r logCPMSepHist, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["logCPMSepHist"]] <- 
  ggplot(featureDF, aes(x = AveLogCPM)) + geom_histogram(bins = 30) + 
  facet_wrap(~dataset, nrow = colRow[2]) +
  xlab("Average log CPM") + thm
plots[["logCPMSepHist"]]
```

### Overlaid histograms

```{r logCPMOverlaidHist}
plots[["logCPMOverlaidHist"]] <- 
  ggplot(featureDF, aes(x = AveLogCPM, group = dataset, fill = dataset)) + 
  geom_histogram(bins = 30, alpha = 0.5, position = "identity") +
  xlab("Average log CPM") + thm
plots[["logCPMOverlaidHist"]]
```

### Density plots

```{r logCPMDensity}
plots[["logCPMDensity"]] <- 
  ggplot(featureDF, aes(x = AveLogCPM, group = dataset, color = dataset)) + 
  geom_line(alpha = 0.5, size = 1.5, stat = "density") +
  xlab("Average log CPM") + thm
plots[["logCPMDensity"]]
```

### Density plots (filled) 

```{r logCPMDensityFilled}
plots[["logCPMDensityFilled"]] <- 
  ggplot(featureDF, aes(x = AveLogCPM, group = dataset, fill = dataset)) + 
  geom_density(alpha = 0.5) +
  xlab("Average log CPM") + thm
plots[["logCPMDensityFilled"]]
```

### Box plots

```{r logCPMBox}
plots[["logCPMBox"]] <- 
  ggplot(featureDF, aes(x = dataset, y = AveLogCPM, color = dataset)) + 
  geom_boxplot(outlier.size = -1, alpha = 0.5) + 
  geom_point(position = position_jitter(width = 0.2), size = 2, alpha = 0.5) + 
  ylab("Average log CPM") + xlab("") + thm
plots[["logCPMBox"]]
```

### Violin plots

```{r logCPMViolin}
plots[["logCPMViolin"]] <- 
  ggplot(featureDF, aes(x = dataset, y = AveLogCPM, fill = dataset)) + 
  geom_violin(alpha = 0.5) + 
  ylab("Average log CPM") + xlab("") + thm
plots[["logCPMViolin"]]
```

### Empirical cumulative distribution function 

```{r logCPMEcdf}
plots[["logCPMEcdf"]] <- 
  ggplot(featureDF, aes(x = AveLogCPM, group = dataset, color = dataset)) + 
  stat_ecdf(size = 1.5, alpha = 0.5) + ylab("Cumulative proportion") + 
  xlab("Average log CPM") + thm
plots[["logCPMEcdf"]]

makeDF(featureDF, "AveLogCPM")
```

## Fraction zeros per sample {.tabset .tabset-pills}

The distribution of the fraction of zeros observed per sample. 

### Separate histograms

```{r fraczeroSampleSepHist, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["fraczeroSampleSepHist"]] <- 
  ggplot(sampleDF, aes(x = Fraczero)) + geom_histogram(bins = 30) + 
  facet_wrap(~dataset, nrow = colRow[2]) +
  xlab("Fraction zeros per sample") + thm
plots[["fraczeroSampleSepHist"]]
```

### Overlaid histograms

```{r fraczeroSampleOverlaidHist}
plots[["fraczeroSampleOverlaidHist"]] <- 
  ggplot(sampleDF, aes(x = Fraczero, group = dataset, fill = dataset)) + 
  geom_histogram(bins = 30, alpha = 0.5, position = "identity") +
  xlab("Fraction zeros per sample") + thm
plots[["fraczeroSampleOverlaidHist"]]
```

### Density plots

```{r fraczeroSampleDensity}
plots[["fraczeroSampleDensity"]] <- 
  ggplot(sampleDF, aes(x = Fraczero, group = dataset, color = dataset)) + 
  geom_line(alpha = 0.5, size = 1.5, stat = "density") +
  xlab("Fraction zeros per sample") + thm
plots[["fraczeroSampleDensity"]]
```

### Density plots (filled) 

```{r fraczeroSampleDensityFilled}
plots[["fraczeroSampleDensityFilled"]] <- 
  ggplot(sampleDF, aes(x = Fraczero, group = dataset, fill = dataset)) + 
  geom_density(alpha = 0.5) +
  xlab("Fraction zeros per sample") + thm
plots[["fraczeroSampleDensityFilled"]]
```

### Box plots

```{r fraczeroSampleBox}
plots[["fraczeroSampleBox"]] <- 
  ggplot(sampleDF, aes(x = dataset, y = Fraczero, color = dataset)) + 
  geom_boxplot(outlier.size = -1, alpha = 0.5) + 
  geom_point(position = position_jitter(width = 0.2), size = 2, alpha = 0.5) + 
  ylab("Fraction zeros per sample") + xlab("") + thm
plots[["fraczeroSampleBox"]]
```

### Violin plots

```{r fraczeroSampleViolin}
plots[["fraczeroSampleViolin"]] <- 
  ggplot(sampleDF, aes(x = dataset, y = Fraczero, fill = dataset)) + 
  geom_violin(alpha = 0.5) + 
  ylab("Fraction zeros per sample") + xlab("") + thm
plots[["fraczeroSampleViolin"]]
```

### Empirical cumulative distribution function 

```{r fraczeroSampleEcdf}
plots[["fraczeroSampleEcdf"]] <- 
  ggplot(sampleDF, aes(x = Fraczero, group = dataset, color = dataset)) + 
  stat_ecdf(size = 1.5, alpha = 0.5) + ylab("Cumulative proportion") + 
  xlab("Fraction zeros per sample") + thm
plots[["fraczeroSampleEcdf"]]

makeDF(sampleDF, "Fraczero")
```

## Fraction zeros per feature {.tabset .tabset-pills}

The distribution of the fraction of zeros observed per variable. 

### Separate histograms

```{r fraczeroFeatureSepHist, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["fraczeroFeatureSepHist"]] <- 
  ggplot(featureDF, aes(x = Fraczero)) + geom_histogram(bins = 30) + 
  facet_wrap(~dataset, nrow = colRow[2]) +
  xlab("Fraction zeros per feature") + thm
plots[["fraczeroFeatureSepHist"]]
```

### Overlaid histograms

```{r fraczeroFeatureOverlaidHist}
plots[["fraczeroFeatureOverlaidHist"]] <- 
  ggplot(featureDF, aes(x = Fraczero, group = dataset, fill = dataset)) + 
  geom_histogram(bins = 30, alpha = 0.5, position = "identity") +
  xlab("Fraction zeros per feature") + thm
plots[["fraczeroFeatureOverlaidHist"]]
```

### Density plots

```{r fraczeroFeatureDensity}
plots[["fraczeroFeatureDensity"]] <- 
  ggplot(featureDF, aes(x = Fraczero, group = dataset, color = dataset)) + 
  geom_line(alpha = 0.5, size = 1.5, stat = "density") +
  xlab("Fraction zeros per feature") + thm
plots[["fraczeroFeatureDensity"]]
```

### Density plots (filled) 

```{r fraczeroFeatureDensityFilled}
plots[["fraczeroFeatureDensityFilled"]] <- 
  ggplot(featureDF, aes(x = Fraczero, group = dataset, fill = dataset)) + 
  geom_density(alpha = 0.5) +
  xlab("Fraction zeros per feature") + thm
plots[["fraczeroFeatureDensityFilled"]]
```

### Box plots

```{r fraczeroFeatureBox}
plots[["fraczeroFeatureBox"]] <- 
  ggplot(featureDF, aes(x = dataset, y = Fraczero, color = dataset)) + 
  geom_boxplot(outlier.size = -1, alpha = 0.5) + 
  geom_point(position = position_jitter(width = 0.2), size = 2, alpha = 0.5) + 
  ylab("Fraction zeros per feature") + xlab("") + thm
plots[["fraczeroFeatureBox"]]
```

### Violin plots

```{r fraczeroFeatureViolin}
plots[["fraczeroFeatureViolin"]] <- 
  ggplot(featureDF, aes(x = dataset, y = Fraczero, fill = dataset)) + 
  geom_violin(alpha = 0.5) + 
  ylab("Fraction zeros per feature") + xlab("") + thm
plots[["fraczeroFeatureViolin"]]
```

### Empirical cumulative distribution function 

```{r fraczeroFeatureEcdf}
plots[["fraczeroFeatureEcdf"]] <- 
  ggplot(featureDF, aes(x = Fraczero, group = dataset, color = dataset)) + 
  stat_ecdf(size = 1.5, alpha = 0.5) + ylab("Cumulative proportion") + 
  xlab("Fraction zeros per feature") + thm
plots[["fraczeroFeatureEcdf"]]

makeDF(featureDF, "Fraczero")
```

## Sample-sample correlations {.tabset .tabset-pills}

The distribution of Spearman correlation coefficients for all pairs of samples, 
calculated on the log-CPM values obtained via the `cpm` function from `edgeR`, 
with a prior.count of 2. If there are more than `r subsampleSize` samples in a
data set, the pairwise correlations between `r subsampleSize` randomly selected
samples are shown.

### Separate histograms

```{r sampleCorrSepHist, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["sampleCorrSepHist"]] <- 
  ggplot(sampleCorrDF, aes(x = Correlation)) + geom_histogram(bins = 30) + 
  facet_wrap(~dataset, nrow = colRow[2]) +
  xlab("Sample-sample correlation") + thm
plots[["sampleCorrSepHist"]]
```

### Overlaid histograms

```{r sampleCorrOverlaidHist}
plots[["sampleCorrOverlaidHist"]] <- 
  ggplot(sampleCorrDF, aes(x = Correlation, group = dataset, fill = dataset)) + 
  geom_histogram(bins = 30, alpha = 0.5, position = "identity") +
  xlab("Sample-sample correlation") + thm
plots[["sampleCorrOverlaidHist"]]
```

### Density plots

```{r sampleCorrDensity}
plots[["sampleCorrDensity"]] <- 
  ggplot(sampleCorrDF, aes(x = Correlation, group = dataset, color = dataset)) + 
  geom_line(alpha = 0.5, size = 1.5, stat = "density") +
  xlab("Sample-sample correlation") + thm
plots[["sampleCorrDensity"]]
```

### Density plots (filled)

```{r sampleCorrDensityFilled}
plots[["sampleCorrDensityFilled"]] <- 
  ggplot(sampleCorrDF, aes(x = Correlation, group = dataset, fill = dataset)) + 
  geom_density(alpha = 0.5) +
  xlab("Sample-sample correlation") + thm
plots[["sampleCorrDensityFilled"]]
```

### Box plots

```{r sampleCorrBox}
plots[["sampleCorrBox"]] <- 
  ggplot(sampleCorrDF, aes(x = dataset, y = Correlation, color = dataset)) + 
  geom_boxplot(outlier.size = -1, alpha = 0.5) + 
  geom_point(position = position_jitter(width = 0.2), size = 2, alpha = 0.5) + 
  ylab("Sample-sample correlation") + xlab("") + thm
plots[["sampleCorrBox"]]
```

### Violin plots

```{r sampleCorrViolin}
plots[["sampleCorrViolin"]] <- 
  ggplot(sampleCorrDF, aes(x = dataset, y = Correlation, fill = dataset)) + 
  geom_violin(alpha = 0.5) + 
  ylab("Sample-sample correlation") + xlab("") + thm
plots[["sampleCorrViolin"]]
```

### Empirical cumulative distribution function 

```{r sampleCorrEcdf}
plots[["sampleCorrEcdf"]] <- 
  ggplot(sampleCorrDF, aes(x = Correlation, group = dataset, color = dataset)) + 
  stat_ecdf(size = 1.5, alpha = 0.5) + ylab("Cumulative proportion") + 
  xlab("Sample-sample correlation") + thm
plots[["sampleCorrEcdf"]]

makeDF(sampleCorrDF, "Correlation")
```

## Variable-variable correlations {.tabset .tabset-pills}

The distribution of Spearman correlation coefficients for all pairs of samples, 
calculated on the log-CPM values obtained via the `cpm` function from `edgeR`, 
with a prior.count of 2. Only non-constant variables are considered, and if 
there are more than `r subsampleSize` such variables in a data set, the pairwise
correlations between `r subsampleSize` randomly selected variables are shown.

### Separate histograms

```{r featureCorrSepHist, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["featureCorrSepHist"]] <- 
  ggplot(featureCorrDF, aes(x = Correlation)) + geom_histogram(bins = 30) + 
  facet_wrap(~dataset, nrow = colRow[2]) +
  xlab("Variable-variable correlation") + thm
plots[["featureCorrSepHist"]]
```

### Overlaid histograms

```{r featureCorrOverlaidHist}
plots[["featureCorrOverlaidHist"]] <- 
  ggplot(featureCorrDF, aes(x = Correlation, group = dataset, fill = dataset)) + 
  geom_histogram(bins = 30, alpha = 0.5, position = "identity") +
  xlab("Variable-variable correlation") + thm
plots[["featureCorrOverlaidHist"]]
```

### Density plots

```{r featureCorrDensity}
plots[["featureCorrDensity"]] <- 
  ggplot(featureCorrDF, aes(x = Correlation, group = dataset, color = dataset)) + 
  geom_line(alpha = 0.5, size = 1.5, stat = "density") +
  xlab("Variable-variable correlation") + thm
plots[["featureCorrDensity"]]
```

### Density plots (filled)

```{r featureCorrDensityFilled}
plots[["featureCorrDensityFilled"]] <- 
  ggplot(featureCorrDF, aes(x = Correlation, group = dataset, fill = dataset)) + 
  geom_density(alpha = 0.5) +
  xlab("Variable-variable correlation") + thm
plots[["featureCorrDensityFilled"]]
```

### Box plots

```{r featureCorrBox}
plots[["featureCorrBox"]] <- 
  ggplot(featureCorrDF, aes(x = dataset, y = Correlation, color = dataset)) + 
  geom_boxplot(outlier.size = -1, alpha = 0.5) + 
  geom_point(position = position_jitter(width = 0.2), size = 2, alpha = 0.5) + 
  ylab("Variable-variable correlation") + xlab("") + thm
plots[["featureCorrBox"]]
```

### Violin plots

```{r featureCorrViolin}
plots[["featureCorrViolin"]] <- 
  ggplot(featureCorrDF, aes(x = dataset, y = Correlation, fill = dataset)) + 
  geom_violin(alpha = 0.5) + 
  ylab("Variable-variable correlation") + xlab("") + thm
plots[["featureCorrViolin"]]
```

### Empirical cumulative distribution function 

```{r featureCorrEcdf}
plots[["featureCorrEcdf"]] <- 
  ggplot(featureCorrDF, aes(x = Correlation, group = dataset, color = dataset)) + 
  stat_ecdf(size = 1.5, alpha = 0.5) + ylab("Cumulative proportion") + 
  xlab("Variable-variable correlation") + thm
plots[["featureCorrEcdf"]]

makeDF(featureCorrDF, "Correlation")
```

## Library size vs fraction zeros {.tabset .tabset-pills}

The association between the total count and the fraction of zeros observed per
sample.

### Separate scatter plots

```{r libsizeFraczeroSepScatter, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["libsizeFraczeroSepScatter"]] <- 
  ggplot(sampleDF, aes(x = Libsize, y = Fraczero)) + geom_point(size = 1) + 
  facet_wrap(~dataset, nrow = colRow[2]) + 
  xlab("Library size") + ylab("Fraction zeros") + thm
plots[["libsizeFraczeroSepScatter"]]
```

### Overlaid scatter plots

```{r libsizeFraczeroOverlaidScatter}
plots[["libsizeFraczeroOverlaidScatter"]] <- 
  ggplot(sampleDF, aes(x = Libsize, y = Fraczero, color = dataset)) + 
  geom_point(size = 1, alpha = 0.5) + 
  geom_line(stat = "smooth", method = "loess", size = 1, alpha = 0.5) + 
  xlab("Library size") + ylab("Fraction zeros") + thm
plots[["libsizeFraczeroOverlaidScatter"]]
```

## Mean expression vs fraction zeros {.tabset .tabset-pills}

The association between the average abundance and the fraction of zeros observed
per variable.

### Separate scatter plots

```{r logCPMFraczeroSepScatter, fig.width = panelSize * colRow[1], fig.height = panelSize * colRow[2]}
plots[["logCPMFraczeroSepScatter"]] <- 
  ggplot(featureDF, aes(x = AveLogCPM, y = Fraczero)) + geom_point(size = 0.75) + 
  facet_wrap(~dataset, nrow = colRow[2]) + 
  xlab("Average log CPM") + ylab("Fraction zeros") + thm
plots[["logCPMFraczeroSepScatter"]]
```

### Overlaid scatter plots

```{r logCPMFraczeroOverlaidScatter}
plots[["logCPMFraczeroOverlaidScatter"]] <- 
  ggplot(featureDF, aes(x = AveLogCPM, y = Fraczero, color = dataset)) + 
  geom_point(size = 0.75, alpha = 0.5) + 
  xlab("Average log CPM") + ylab("Fraction zeros") + thm
plots[["logCPMFraczeroOverlaidScatter"]]
```


```{r, include = FALSE}
if (savePlots) 
  saveRDS(plots, 
          file = paste0(outputDir, "/",
                        tools::file_path_sans_ext(basename(outputFile)), "_ggplots.rds"))
```

## Session info

```{r}
sessionInfo()
```

## References

